File: solutions/php/meetup/Meetup.php
<?php

declare(strict_types=1);

function meetup_day(int $year, int $month, string $which, string $weekday): DateTimeImmutable
{
    $day_num = array(
        "Monday"    =>1,
        "Tuesday"   =>2,
        "Wednesday" =>3,
        "Thursday"  =>4,
        "Friday"    =>5,
        "Saturday"  =>6,
        "Sunday"    =>7
    );

    $first_day = array(
        "first"  =>1,
        "second" =>8,
        "third"  =>15,
        "fourth" =>22,
        "teenth" =>13
    );

    if ($which != "last") {
        $day = $first_day[$which];  // start day of requested week
        $dir = 1;
    } else {
        $day = date("t", strtotime("$year-$month-1")); // last day
        $dir = -1;
    }

    $guess = new DateTimeImmutable("$year-$month-$day");
    $offset = ($day_num[$weekday] - $guess->format("w") + 7 * $dir) % 7;
    return $guess->modify("$offset days");
}

File: solutions/erlang/meetup/src/meetup.erl
-module(meetup).

-export([meetup/4]).

meetup(Year, Month, DayOfWeek, last) ->
    LastDay = calendar:last_day_of_the_month(Year, Month),
    adjust_day_in_week(DayOfWeek, {Year, Month, LastDay}, -1);
meetup(Year, Month, DayOfWeek, Week) ->
    adjust_day_in_week(DayOfWeek, {Year, Month, first_day(Week)}, 1).

adjust_day_in_week(DayOfWeek, Guess, Dir) ->
    Offset = (day_num(DayOfWeek) - calendar:day_of_the_week(Guess) + 7 * Dir) rem 7,
    calendar:gregorian_days_to_date(calendar:date_to_gregorian_days(Guess) + Offset).

day_num(monday) -> 1;
day_num(tuesday) -> 2;
day_num(wednesday) -> 3;
day_num(thursday) -> 4;
day_num(friday) -> 5;
day_num(saturday) -> 6;
day_num(sunday) -> 7.

first_day(first) -> 1;
first_day(second) -> 8;
first_day(third) -> 15;
first_day(fourth) -> 22;
first_day(teenth) -> 13.
